lib/github/parse_owner.ex

```diff
@@ -0,0 +1,104 @@
+require Logger
+
+defmodule BorsNG.CodeOwners do
+  @type tjson :: map
+  @type t :: %BorsNG.CodeOwners{
+          patterns: [BorsNG.FilePattern]
+        }
+  defstruct(patterns: [])
+end
+
+defmodule BorsNG.FilePattern do
+  @type tjson :: map
+  @type t :: %BorsNG.FilePattern{
+          file_pattern: bitstring,
+          approvers: [bitstring]
+        }
+  defstruct(
+    file_pattern: "",
+    approvers: {}
+  )
+end
+
+defmodule BorsNG.CodeOwnerParser do
+  # Returns a list of lists
+  # Items in the inner lists are joined by an OR statement
+  # Items in the the outer list are joined by an AND statement
+  # [ [A], [A, B], [A, C] -> A and (A or B) and (A or C)
+  @spec list_required_reviews(%BorsNG.CodeOwners{}, [%BorsNG.GitHub.File{}]) :: [[bitstring]]
+  def list_required_reviews(code_owners, files) do
+    Logger.debug("Code Owners: #{inspect(code_owners)}")
+    Logger.debug("Files modified: #{inspect(files)}")
+
+    required_reviewers =
+      Enum.map(files, fn x ->
+        pats =
+          Enum.map(code_owners.patterns, fn owner ->
+            cond do
+              # glob matches has an extra '/' in it, don't match on it
+             :glob.matches(x.filename, owner.file_pattern) && !:glob.matches(x.filename, owner.file_pattern <> "/*")  ->
+                owner.approvers
+             String.starts_with?(x.filename, owner.file_pattern) ->
+              owner.approvers
+             true ->
+              nil # if unknown fall through
+            end
+          end)
```

Comment:
This looks like a case where you *should* use `flat_map`:

```elixir
          Enum.map(code_owners.patterns, fn owner ->
            cond do
              # glob matches has an extra '/' in it, don't match on it
             :glob.matches(x.filename, owner.file_pattern) && !:glob.matches(x.filename, owner.file_pattern <> "/*")  ->
                [owner.approvers]
             String.starts_with?(x.filename, owner.file_pattern) ->
              [owner.approvers]
             true ->
              [] # if unknown fall through
            end
          end)
```

(the difference, of course, is that in this function, the same analysis is responsible both for filtering and for mapping)

Reply:
In this case I don't think so.

I need to do some odd filtering.  A file can only have one required reviewer and the last match takes precedence.  So I need to keep the reviewer list together until later.

I've added some comments discussing this.